src/finufft.cpp

```diff
@@ -821,12 +826,10 @@ int FINUFFT_SETPTS(FINUFFT_PLAN p, BIGINT nj, FLT* xj, FLT* yj, FLT* zj,
       fprintf(stderr, "[%s t3] fwBatch would be bigger than MAX_NF, not attempting malloc!\n",__func__);
       return ERR_MAXNALLOC;
     }
-#pragma omp critical
-    {
-      if (p->fwBatch)
-        FFTW_FR(p->fwBatch);
-      p->fwBatch = FFTW_ALLOC_CPX(p->nf * p->batchSize); // maybe big workspace
-    }
+    if (p->fwBatch)
+      FFTW_FR(p->fwBatch);
```

Comment:
no sure why no mutex here, but the free below is mutexed?

Reply:
@ahbarnett For this, Robert and I discussed in issue #353.

For `FFTW_ALLOC_CPX` and `FFTW_FR`, the actual fftw memory alloc and memory free are calling the system's alloc and free, see [https://github.com/FFTW/fftw3/blob/master/kernel/kalloc.c](https://github.com/FFTW/fftw3/blob/master/kernel/kalloc.c) which should be thread safe in the modern standard.

For `FFTW_DE`, the destroy logic is [https://github.com/FFTW/fftw3/blob/38ea230e25e69e7a3f35b957b815bac4f9aa22b0/api/apiplan.c#L194](https://github.com/FFTW/fftw3/blob/38ea230e25e69e7a3f35b957b815bac4f9aa22b0/api/apiplan.c#L194), it will have some thread issues if not locked. Also other fftw calls are not thread-safe, except that `fftw_execute` is thread-safe.

So Robert puts all fftw calls in mutex except `fftw_execute`, `fftw_alloc` and `fftw_free`.